Java/jni/native_glue.cc

```diff
@@ -17,9 +17,134 @@
 //
 
 #include "native_glue.hh"
+#include <queue>
+#include <new>
 
 using namespace litecore;
 using namespace litecore::jni;
+using namespace std;
+
+namespace litecore { namespace jni {
+        void UTF8ToModifiedUTF8(const char *input, char *output);
+        void ModifiedUTF8ToUTF8(char* input);
+        void ModifiedUTF8CharToUTF8(char* input);
+    }
+}
+
+void litecore::jni::UTF8ToModifiedUTF8(const char *input, char *output) {
+    char c = input[0];
+    char c1 = input[1] & 0x3F;
+    char c2 = input[2] & 0x3F;
+    char c3 = input[3] & 0x3F;
+
+    int unicodePoint = ((c & 0x07) << 18) | (c1 << 12) | (c2 << 6) | c3;
+    unicodePoint -= 0x10000;
+
+    int surrogates[2] { 0xD800 | (unicodePoint >> 10), 0xDC00 | (unicodePoint & 0x3FF) };
+
+    output[0] = surrogates[0]>>12 & 0x0F | 0xE0;
+    output[1] = surrogates[0]>>6 & 0x3F | 0x80;
+    output[2] = surrogates[0] & 0x3F | 0x80;
+    output[3] = surrogates[1]>>12 & 0x0F | 0xE0;
+    output[4] = surrogates[1]>>6 & 0x3F | 0x80;
+    output[5] = surrogates[1] & 0x3F | 0x80;
+}
+
+size_t litecore::jni::UTF8ToModifiedUTF8(const char* input, const char** output, size_t len){
+    // https://github.com/android-ndk/ndk/issues/283
+    size_t extraBytes = 0;
+    const auto unsignedInput = (const uint8_t *)input;
+
+    // Need to figure out the actual length since each
+    // 4-bytes of real UTF-8 is 6 bytes of modified UTF-8
+    // but convert the bytes while we are at it
+    for(size_t i = 0; i < len; i++) {
+        if(unsignedInput[i] >= 0xF0) {
+            // 0xF0 and above marks 4-byte sequences
+            extraBytes += 2;
+
+            // 3 + 1 from next loop iteration = 4 bytes advance
+            i += 3;
+        }
+    }
+
+    if(extraBytes == 0) {
+        // No modifications necessary
+        *output = input;
+        return len;
+    }
+
+    size_t newStrLen = len + extraBytes;
+    char* newBytes = (char *)malloc(newStrLen);
+    if(newBytes == nullptr) {
+        *output = nullptr;
+        return 0;
```

Comment:
0 is a legal return value (for an empty input string.) Maybe return -1 on error (and change type to `ssize_t`?)